### PR TITLE
hle: return enosys on unpresented syscalls

### LINKED ISSUE #4071: [BLES02237] Unimplemented syscall 462
Hi rpcs3 developers,

rpcs3 [435ea8d55379a887442916ff309c6f1e28fc9652](https://github.com/RPCS3/rpcs3/commit/435ea8d55379a887442916ff309c6f1e28fc9652) is pretty cool about binary translation through LLVM! but unimplemented syscall 462:

```
U {PPU[0x1000000] Thread (main_thread) [0x028c10e4]} HLE TODO: Unimplemented syscall syscall_462 -> CELL_OK
U {PPU[0x1000000] Thread (main_thread) [0x028d0e64]} sys_prx TODO: _sys_prx_register_module()
U {PPU[0x1000000] Thread (main_thread) [0x028d10bc]} sys_prx TODO: _sys_prx_register_library(library=*0x28d36a0)
...
U {PPU[0x1000000] Thread (main_thread) [0x028dcd8c]} HLE TODO: Unimplemented syscall sys_ss_access_control_engine -> CELL_OK
...
U {PPU[0x1000000] Thread (main_thread) [0x028d161c]} sys_prx TODO: _sys_prx_register_module()
E {PPU[0x1000000] Thread (main_thread) [0x00134758]} sys_mmapper: sys_mmapper_allocate_address(size=0x20000000, flags=0x200, alignment=0x10000000, alloc_addr=*0xd00204e0)
E {PPU[0x1000000] Thread (main_thread) [0x00134778]} sys_mmapper: sys_mmapper_allocate_address(size=0x20000000, flags=0x200, alignment=0x10000000, alloc_addr=*0xd00204e0)
E {PPU[0x1000000] Thread (main_thread) [0x028c7948]} 'sys_mmapper_map_shared_memory' failed with 0x80010010 : CELL_EALIGN [1]
F {PPU[0x1000000] Thread (main_thread) [0x028c792c]} St13runtime_error thrown: Unregistered PPU function
```

So I checked the [COMPATIBILITY](https://rpcs3.net/compatibility?g=pro+evolution) for [BLES02237](https://forums.rpcs3.net/thread-185614.html), then I realized :)

But how to fix? The [patch](https://github.com/RPCS3/rpcs3/pull/2737/) is not related to this bug.
I want to dig into rpcs3, please teaching me, thanks a lot!

Regards,
Leslie Zhai - https://reviews.llvm.org/p/xiangzhai/

### PR DESCRIPTION
**Pull request by elad335**
Partially fixes #4071

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 1]
```diff
@@ -274,9 +283,7 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_isolated_spu_set_spu_cfg)     //238 (0x0EE)  ROOT
 	null_func,//BIND_FUNC(sys_isolated_spu_get_spu_cfg)     //239 (0x0EF)  ROOT
 	null_func,//BIND_FUNC(sys_isolated_spu_read_puint_mb)   //240 (0x0F0)  ROOT
-	null_func,                                              //241 (0x0F1)  ROOT  UNS
-	null_func,                                              //242 (0x0F2)  ROOT  UNS
-	null_func,                                              //243 (0x0F3)  ROOT  UNS
+	UNS, UNS, UNS,                                          //243 (0x0F3)  ROOT  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should the comment be changed to `//241-243 (0x0F1-0x0F3)  ROOT  UNS` ?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 2]
```diff
@@ -289,18 +296,14 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
...
-	null_func, null_func, null_func, null_func, null_func,  //274  UNS
-	null_func, null_func, null_func, null_func, null_func,  //279  UNS
-	null_func, null_func, null_func, null_func, null_func,  //284  UNS
-	null_func, null_func, null_func, null_func, null_func,  //289  UNS
-	null_func, null_func, null_func, null_func, null_func,  //294  UNS
-	null_func, null_func, null_func, null_func, null_func,  //299  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,            //269  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //279  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //289  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //299  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 3]
```diff
@@ -289,18 +296,14 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
...
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above 4 comments be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 4]
```diff
@@ -362,13 +365,12 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	BIND_FUNC(sys_mmapper_allocate_shared_memory_from_container),//362 (0x16A)
 	null_func,//BIND_FUNC(sys_mmapper_...)                  //363 (0x16B)
 	null_func,//BIND_FUNC(sys_mmapper_...)                  //364 (0x16C)
-	null_func,                                              //365 (0x16D)  UNS
-	null_func,                                              //366 (0x16E)  UNS
+	UNS, UNS,                                               //366 (0x16E)  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 5]
```diff
@@ -414,13 +416,10 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
...
-	null_func, null_func, null_func, null_func, null_func,  //424  UNS
-	null_func, null_func, null_func, null_func, null_func,  //429  UNS
-	null_func, null_func, null_func, null_func, null_func,  //434  UNS
-	null_func, null_func, null_func, null_func, null_func,  //439  UNS
-	null_func, null_func, null_func, null_func, null_func,  //444  UNS
-	null_func, null_func, null_func, null_func, null_func,  //449  UNS
+	UNS, UNS, UNS, UNS,                                     //419  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //429  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //439  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //449  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above 4 comments be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 6]
```diff
@@ -498,10 +497,7 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_config_io_event)              //523 (0x20B)
 	null_func,//BIND_FUNC(sys_config_...)                   //524 (0x20C)
 	null_func,//BIND_FUNC(sys_config_...)                   //525 (0x20D)
-	null_func,                                              //526 (0x20E)  UNS
-	null_func,                                              //527 (0x20F)  UNS
-	null_func,                                              //528 (0x210)  UNS
-	null_func,                                              //529 (0x211)  UNS
+	UNS, UNS, UNS, UNS,                                     //529 (0x211)  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 7]
```diff
@@ -600,11 +596,10 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_io_buffer_destroy)            //625 (0x271)
 	null_func,//BIND_FUNC(sys_io_buffer_allocate)           //626 (0x272)
 	null_func,//BIND_FUNC(sys_io_buffer_free)               //627 (0x273)
-	null_func,                                              //628 (0x274)  UNS
-	null_func,                                              //629 (0x275)  UNS
+	UNS, UNS,                                               //629 (0x275)  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 8]
```diff
@@ -627,8 +622,7 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_rsxaudio_stop_process)        //658 (0x292)
 	null_func,//BIND_FUNC(sys_rsxaudio_)                    //659 (0x293)
 
-	null_func, null_func, null_func, null_func, null_func,  //664  UNS
-	null_func,                                              //665  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS,                           //665  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 9]
```diff
@@ -692,21 +686,14 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
...
-	null_func, null_func, null_func, null_func, null_func,  //794  UNS
-	null_func, null_func, null_func, null_func, null_func,  //799  UNS
+	UNS, UNS, UNS,                                          //729  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //739  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //749  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //759  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //769  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //779  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //789  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //799  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above 8 comments be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 10]
```diff
@@ -757,9 +744,9 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	BIND_FUNC(sys_fs_mapped_free),                          //846 (0x34E)
 	BIND_FUNC(sys_fs_truncate2),                            //847 (0x34F)
 
-	null_func, null_func,                                   //849  UNS
-	null_func, null_func, null_func, null_func, null_func,  //854  UNS
-	null_func, null_func, null_func, null_func, null_func,  //859  UNS
+	UNS, UNS,                                               //849  UNS
+	UNS, UNS, UNS, UNS, UNS,                                //854  UNS
+	UNS, UNS, UNS, UNS, UNS,                                //859  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 11]
```diff
@@ -787,9 +774,7 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_deci3_send)                   //883 (0x373)
 	null_func,//BIND_FUNC(sys_deci3_receive)                //884 (0x374)
 	null_func,//BIND_FUNC(sys_deci3_open2)                  //885 (0x375)
-	null_func,                                              //886 (0x376)  UNS
-	null_func,                                              //887 (0x377)  UNS
-	null_func,                                              //888 (0x378)  UNS
+	UNS, UNS, UNS,                                          //888 (0x378)  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above comment be changed to with syscall number range?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 12]
```diff
@@ -892,13 +877,10 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
...
-	null_func, null_func, null_func, null_func, null_func,  //999  UNS
-	null_func, null_func, null_func, null_func, null_func,  //1004  UNS
-	null_func, null_func, null_func, null_func, null_func,  //1009  UNS
-	null_func, null_func, null_func, null_func, null_func,  //1014  UNS
-	null_func, null_func, null_func, null_func, null_func,  //1019  UNS
-	null_func, null_func, null_func, null_func,             //1023  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //999  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //1009  UNS
+	UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS, UNS,       //1019  UNS
+	UNS, UNS, UNS, UNS,                                     //1023  UNS
```
Comment (mzji) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Should above 4 comments be changed to with syscall number range?

**Comment by elad335** [comment]
@mzji It would look ugly though

**Comment by Nekotekina** [comment]
My opinion: just return ENOSYS in null_func, there is no need to touch the table and duplicate.

**Comment by elad335** [comment]
lmao

File: rpcs3/Emu/Cell/PPUFunction.cpp [hunk 13]
```diff
@@ -570,7 +570,7 @@ extern std::string ppu_get_syscall_name(u64 code)
 	case 985: return "sys_dbg_get_console_type";
 	}
 
-	return fmt::format("syscall_%llu", code);
+	return fmt::format("%llu", code);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPUFunction.cpp]: These strings are used in LLVM, you shouldn't just edit them.

Comment (elad335) [reply rpcs3/Emu/Cell/PPUFunction.cpp]: hmm ok

**Comment by sftlbcn** [comment]
breaks tlou

> U {PPU[0x1000000] Thread (main_thread) [0x01533d8c]} HLE TODO: Unimplemented syscall sys_ss_access_control_engine -> ENOSYS
U {PPU[0x1000000] Thread (main_thread) [0x01533d24]} HLE TODO: Unimplemented syscall sys_ss_access_control_engine -> ENOSYS
U {PPU[0x1000000] Thread (main_thread) [0x0152f860]} sys_prx TODO: _sys_prx_get_module_list(flags=2, pInfo=*0xd003fd88)
U {PPU[0x1000000] Thread (main_thread) [0x0152f860]} sys_prx TODO: _sys_prx_get_module_list(flags=2, pInfo=*0xd00405f8)
E {PPU[0x1000000] Thread (main_thread) [0x0152f780]} 'sys_lwmutex_lock' failed with 0x80010005 : CELL_ESRCH [1]
F {PPU[0x1000000] Thread (main_thread) [0x0091832c]} class std::runtime_error thrown: Trap!
(in file Emu\Cell\PPUInterpreter.cpp:3163)

**Comment by elad335** [comment]
so I'll just do it in the safe way

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 14]
```diff
@@ -49,11 +49,20 @@ static bool null_func(ppu_thread& ppu)
...
+	ppu.cia += 4;
+	return false;
+}
+
 std::array<ppu_function_t, 1024> g_ppu_syscall_table{};
 
 // UNS = Unused
 // ROOT = Root
 // DBG = Debug
+// DEX..DECR = 	Unavailable on retail consoles
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Why you don't provide your tests? Or why you disabled sys_trace?

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/lv2.cpp]: @Nekotekina http://www.psdevwiki.com/ps3/LV2_Functions_and_Syscalls I've actually read it before opening the pr, sys_trace is debug only

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/lv2.cpp]: I've sent you the testcase on discord

Comment (Nekotekina) [reply rpcs3/Emu/Cell/lv2/lv2.cpp]: This wiki is very inexact source of information.

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 15]
```diff
@@ -118,8 +127,8 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	BIND_FUNC(sys_trace_drain),                             //65  (0x041)
 	BIND_FUNC(sys_trace_attach_process),                    //66  (0x042)
 	BIND_FUNC(sys_trace_allocate_buffer),                   //67  (0x043)
-	BIND_FUNC(sys_trace_free_buffer),                       //68  (0x044)
-	BIND_FUNC(sys_trace_create2),                           //69  (0x045)
+	BIND_FUNC(sys_trace_free_buffer),            		    //68  (0x044)
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Wild tabs appear

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 16]
```diff
@@ -432,17 +431,17 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_overlay_get_sdk_version)      //456 (0x1C8)
 	null_func,//BIND_FUNC(sys_overlay_get_module_dbg_info)  //457 (0x1C9)
 	null_func,//BIND_FUNC(sys_overlay_get_module_dbg_info)  //458 (0x1CA)
-	null_func,                                              //459 (0x1CB)  UNS
-	null_func,//BIND_FUNC(sys_prx_dbg_get_module_id_list)   //460 (0x1CC)  ROOT
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Leave all named functions as null

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 17]
```diff
@@ -664,7 +658,7 @@ const std::array<ppu_function_t, 1024> s_ppu_syscall_table
 	null_func,//BIND_FUNC(sys_...)                          //696 (0x2B8)  ROOT
 	null_func,//BIND_FUNC(sys_...)                          //697 (0x2B9)  UNS
 	null_func,//BIND_FUNC(sys_...)                          //698 (0x2BA)  UNS
-	null_func,//BIND_FUNC(sys_bdemu_send_command)           //699 (0x2BB)
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp]: Another one, definitely exists in retail kernel. I won't mention them all, just don't touch anything.

**Pull request merged**